### PR TITLE
[SPARK-4072][Core]Display Streaming blocks in Streaming UI

### PR DESCRIPTION
**Pull request by zsxwing**
Replace #6634

This PR adds `SparkListenerBlockUpdated` to SparkListener so that it can monitor all block update infos that are sent to `BlockManagerMasaterEndpoint`, and also add new tables in the Storage tab to display the stream block infos.

screen shot 2015-07-01 at 5 19 46 pm

### CODE REVIEW

**Comment by zsxwing** [comment]
cc @tdas

**Comment by tdas** [comment]
Awesome! Screenshot looking pretty good. Here are some nit picky comments.

- RDD, Receiver Blocks, these titles are same size as "Storage" at the top. Should be smaller. 
- Address / Exec ID format is little confusing. Might be a good idea to ignore the exec ID. If some one needs to relate it to the executor, they can use the hostname to find it themselves on the executors page.
- How will it look if one of the 2 copies of the blocks fall off from the executor?

Related
- We should not drop blocks from the UI, because the Storage page is meant to be a true representation of the what is going on in the cluster. Instead we could make multiple collapsible sections, one for each input stream 
- Let's not put the broadcast blocks in the page yet, because we are not even sure whether its worth putting those blocks, whether people even need to debug that. And whether showing the individual pieces is worth it or not. Lets keep those questions outside the scope of this PR, and address them in a separate PR.

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+              blockId,
+              newStorageLevel,
+              memSize,
+              diskSize,
+              externalBlockStoreSize,
+              newLocations))
+        } else {
+          // If isValid is not true, it means we should drop the block.
+          blocksInBlockManager -= blockId
+          removeBlockFromBlockManager(blockId, blockManagerId)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: > How will it look if one of the 2 copies of the blocks fall off from the executor?

`2x` will become `1x`. When one of 2 copies is removed, an update info with `StorageLevel.None` will be sent. Then `removeBlockFromBlockManager` will create a new `StorageLevel` with `replication = 1` for this block.

Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: That's cool then. Will take a more detailed look at the code soon.

**Comment by zsxwing** [comment]
New screenshot

screen shot 2015-06-06 at 10 30 12 pm

**Comment by zsxwing** [comment]
retest this please

**Comment by tdas** [comment]
@JoshRosen Can you take a look at this PR, especially the SparkListener updates.

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala [hunk 2]
```diff
@@ -98,6 +99,9 @@ case class SparkListenerExecutorAdded(time: Long, executorId: String, executorIn
 case class SparkListenerExecutorRemoved(time: Long, executorId: String, reason: String)
   extends SparkListenerEvent
 
+@DeveloperApi
+case class SparkListenerBlockUpdated(updateBlockInfo: UpdateBlockInfo) extends SparkListenerEvent
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: Isnt UpdateBlockInfo private[spark]? Then its not right that a public class refer to an internal class that people will not be able to refer to.

File: core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala [hunk 3]
```diff
@@ -30,11 +30,32 @@ private[ui] class StoragePage(parent: StorageTab) extends WebUIPage("") {
...
+    val streamBlocks = blocks.filter(_.blockId.isInstanceOf[StreamBlockId])
+    val broadcastBlocks = blocks.filter(_.blockId.isInstanceOf[BroadcastBlockId])
+    val content = rddTable ++ blockTable("Receiver Block", streamBlocks)
     UIUtils.headerSparkPage("Storage", content, parent)
   }
 
+  private def rddTable: Seq[Node] = {
+    val rdds = listener.rddInfoList
+    <div>
+      <h4>RDD</h4>
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala]: RDD --> RDDs

File: core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala [hunk 4]
```diff
@@ -30,11 +30,32 @@ private[ui] class StoragePage(parent: StorageTab) extends WebUIPage("") {
   private val listener = parent.listener
 
   def render(request: HttpServletRequest): Seq[Node] = {
-    val rdds = listener.rddInfoList
-    val content = UIUtils.listingTable(rddHeader, rddRow, rdds, id = Some("storage-by-rdd-table"))
+    val blocks = listener.allBlocks.sortBy(_.blockId.toString)
+    val streamBlocks = blocks.filter(_.blockId.isInstanceOf[StreamBlockId])
+    val broadcastBlocks = blocks.filter(_.blockId.isInstanceOf[BroadcastBlockId])
+    val content = rddTable ++ blockTable("Receiver Block", streamBlocks)
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala]: Receiver Block --> Receiver Blocks

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        }
+      }
+    }
+  }
+
+  override def onBlockManagerAdded(blockManagerAdded: SparkListenerBlockManagerAdded): Unit =
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: nit: please add a `{` in this line.

File: core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala (hunk 2 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/SparkListener.scala]: Right. Just added a new develop class `BlockUpdatedInfo`. And I also updated `JsonProtocol` to support the new `SparkListenerBlockUpdated`.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by JoshRosen** [comment]
I've noticed that you're not persisting the SparkListenerBlockUpdated events in the event logging listener.  Since there might be _tons_ of these events, I can understand why we might not want to persist them in the event log, similar to how we don't persist ExecutorMetricsUpdate events.  If this is intentional, I think we should add an override and comment to EventLoggingListener to make this more explicit: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala#L201

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 6]
```diff
@@ -848,6 +908,60 @@ private[spark] object JsonProtocol {
...
+  def blockIdFromJson(json: JValue): BlockId = {
+    val rddBlockId = Utils.getFormattedClassName(RDDBlockId)
+    val shuffleBlockId = Utils.getFormattedClassName(ShuffleBlockId)
+    val shuffleDataBlockId = Utils.getFormattedClassName(ShuffleDataBlockId)
+    val shuffleIndexBlockId = Utils.getFormattedClassName(ShuffleIndexBlockId)
+    val broadcastBlockId = Utils.getFormattedClassName(BroadcastBlockId)
+    val taskResultBlockId = Utils.getFormattedClassName(TaskResultBlockId)
+    val streamBlockId = Utils.getFormattedClassName(StreamBlockId)
+
+    (json \ "Class Name").extract[String] match {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: What happens if we get an unknown class name here?  That situation might occur if we're parsing a newer event log using an older version of the Spark HistoryServer.  This point might be moot if we're not persisting these events to the log (see my earlier comment about EventLoggingListener).  If this is safe due to lack of persistence, maybe we should just leave a note here explaining that.

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    }
+    val blockManagerId = blockUpdated.blockUpdatedInfo.blockManagerId
+    val storageLevel = blockUpdated.blockUpdatedInfo.storageLevel
+    val memSize = blockUpdated.blockUpdatedInfo.memSize
+    val diskSize = blockUpdated.blockUpdatedInfo.diskSize
+    val externalBlockStoreSize = blockUpdated.blockUpdatedInfo.externalBlockStoreSize
+
+    synchronized {
+      // Drop the update info if the block manager is not registered
+      blockManagers.get(blockManagerId) foreach { blocksInBlockManager =>
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: Style nit: should have a `.` before `foreach`

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    synchronized {
+      blockManagers.put(blockManagerAdded.blockManagerId, mutable.HashSet())
+    }
+  }
+
+  override def onBlockManagerRemoved(
+      blockManagerRemoved: SparkListenerBlockManagerRemoved): Unit = {
+    val blockManagerId = blockManagerRemoved.blockManagerId
+    synchronized {
+      blockManagers.remove(blockManagerId) foreach { blockIds =>
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: Same style nit here regarding infix notation.

File: core/src/main/scala/org/apache/spark/storage/BlockUpdatedInfo.scala [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+package org.apache.spark.storage
+
+import org.apache.spark.annotation.DeveloperApi
+import org.apache.spark.storage.BlockManagerMessages.UpdateBlockInfo
+
+/**
+ * :: DeveloperApi ::
+ * Stores information about a block status in a block manager.
+ */
+@DeveloperApi
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockUpdatedInfo.scala]: I guess that this has to be `@DeveloperApi` because it's referenced by a SparkListenerEvent?

File: core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala [hunk 10]
```diff
@@ -30,11 +30,32 @@ private[ui] class StoragePage(parent: StorageTab) extends WebUIPage("") {
   private val listener = parent.listener
 
   def render(request: HttpServletRequest): Seq[Node] = {
-    val rdds = listener.rddInfoList
-    val content = UIUtils.listingTable(rddHeader, rddRow, rdds, id = Some("storage-by-rdd-table"))
+    val blocks = listener.allBlocks.sortBy(_.blockId.toString)
+    val streamBlocks = blocks.filter(_.blockId.isInstanceOf[StreamBlockId])
+    val broadcastBlocks = blocks.filter(_.blockId.isInstanceOf[BroadcastBlockId])
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala]: Won't this list always be empty because the listener only tracks information for StreamBlockIds?

File: core/src/main/scala/org/apache/spark/ui/storage/StorageTab.scala [hunk 11]
```diff
@@ -39,7 +39,9 @@ private[ui] class StorageTab(parent: SparkUI) extends SparkUITab(parent, "storag
  * This class is thread-safe (unlike JobProgressListener)
  */
 @DeveloperApi
-class StorageListener(storageStatusListener: StorageStatusListener) extends SparkListener {
+class StorageListener(storageStatusListener: StorageStatusListener)
+  extends BlockStatusListener with SparkListener {
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/ui/storage/StorageTab.scala]: AFAIK you can just extend `BlockStatusListener` here, which will transitively pull in the `SparkListener` interface.

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 12]
```diff
@@ -407,6 +414,52 @@ private[spark] object JsonProtocol {
...
+      ("Class Name" -> Utils.getFormattedClassName(blockId)) ~
+      ("Task ID" -> taskId)
+    case StreamBlockId(streamId, uniqueId) =>
+      ("Class Name" -> Utils.getFormattedClassName(blockId)) ~
+      ("Stream ID" -> streamId) ~
+      ("Unique ID" -> uniqueId)
+    // These aren't used, but keeps compiler happy
+    case TempLocalBlockId(_) => JNothing
+    case TempShuffleBlockId(_) => JNothing
+    case TestBlockId(_) => JNothing
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: Should we maybe add a catch-call case here to map unknown block id types to JNothing?

File: core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala [hunk 13]
```diff
@@ -64,4 +85,36 @@ private[ui] class StoragePage(parent: StorageTab) extends WebUIPage("") {
     </tr>
     // scalastyle:on
   }
+
+  private def blockHeader = Seq(
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala]: This could be a `val`, right?

File: core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala (hunk 10 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/ui/storage/StoragePage.scala]: Right. Forgot to remove it when I removed broadcast things from this page.

File: core/src/main/scala/org/apache/spark/storage/BlockUpdatedInfo.scala (hunk 9 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/storage/BlockUpdatedInfo.scala]: Right.

**Comment by zsxwing** [comment]
> I've noticed that you're not persisting the SparkListenerBlockUpdated events in the event logging listener. Since there might be tons of these events, I can understand why we might not want to persist them in the event log, similar to how we don't persist ExecutorMetricsUpdate events. If this is intentional, I think we should add an override and comment to EventLoggingListener to make this more explicit: https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala#L201

@JoshRosen Forgot to add to `EventLoggingListener`. Do you think if people want to see blocks in the history server? They are the final blocks when the driver exits. Right? I think most of people are only interested in the ongoing blocks when the application runs.

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala (hunk 6 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: > That situation might occur if we're parsing a newer event log using an older version of the Spark HistoryServer.

Good point. But actually, I think JsonProtocol only supports backward-compatibility. `JsonProtocol.sparkEventFromJson` does not handle unknown class names either. Because `ReplayListenerBus` will throw an exception if parsing a json event log unsuccessfully, it cannotsupport forward-compatibility.

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: BTW. This will throw `scala.MatchError` for an unknown class name.

Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: It looks like ReplayListenerBus ties to log exceptions except when it receives an IOException (which might be fatal from its perspective).  AFAIK it will try to keep going when encountering unknown events, but I don't know that there's a test for this. I'll investigate.

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala (hunk 12 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: The disadvantage of adding a `case _` is if we add a new BlockId type but forget to add it here, we won't get the compiler warning: `match may not be exhaustive. It would fail on the following input: ...`

**Comment by zsxwing** [comment]
If we decide not to add `SparkListenerBlockUpdated` to `EventLoggingListener`, I will revert the changes to JsonProtocol.

**Comment by JoshRosen** [comment]
For the HistoryServer, it would only really be the final aggregated state that we care about, except the problem is that the event log would store the entire event stream used to construct that state, which could be huge. There might be hacks / tricks that we could come up with to persist the aggregated block info but we might want to consider them separately as part of a larger discussion of HistoryServer designs for long-running applications.

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala (hunk 6 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala]: `ReplayListenerBus` will log exceptions except IOException. But it also skips the rest content in the file. BTW, `scala.MatchError` is a `RuntimeException`.

**Comment by zsxwing** [comment]
OK. Let's don't log SparkListenerBlockUpdated. I have reverted JsonProtocol.

**Comment by sujkh85** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-4072][Core]Display Streaming blocks in Streaming UI (#6672)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by zsxwing** [comment]
retest this please

File: core/src/test/scala/org/apache/spark/storage/BlockStatusListenerSuite.scala [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+    listener.onBlockUpdated(SparkListenerBlockUpdated(
+      BlockUpdatedInfo(
+        blockManagerId,
+        StreamBlockId(0, 100),
+        StorageLevel.MEMORY_AND_DISK,
+        memSize = 100,
+        diskSize = 100,
+        externalBlockStoreSize = 0)))
+    val blocks = listener.allBlocks
+    assert(blocks.size === 1)
```
Comment (tdas) [inline core/src/test/scala/org/apache/spark/storage/BlockStatusListenerSuite.scala]: Can you add more comments on what each of these sets of asserts test? Since it all inside "basic functions" its hard to understand what are actually being tested. Actually this one unit test " basic functions" is fairly long. If it doesnt affect efficiency much, I suggest breaking it into multiple unit tests. 

TL;DR either add intermediate comments OR break it down into smaller parts.

**Comment by tdas** [comment]
Can you put a final screenshot of what it looks like.

**Comment by tdas** [comment]
Other than these small comments, LGTM as is.

File: core/src/test/scala/org/apache/spark/storage/BlockStatusListenerSuite.scala (hunk 14 above)
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/storage/BlockStatusListenerSuite.scala]: Added more comments and rewrote the asserts for readability.

Breaking into small unit tests needs many duplicate codes to construct the test cases. So I just added comments.

**Comment by zsxwing** [comment]
I replaced the screenshot with a new one in the description of this PR.

**Comment by tdas** [comment]
@JoshRosen Any more thoughts on this PR?

File: core/src/main/scala/org/apache/spark/storage/BlockUpdatedInfo.scala [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+case class BlockUpdatedInfo(
+    blockManagerId: BlockManagerId,
+    blockId: BlockId,
+    storageLevel: StorageLevel,
+    memSize: Long,
+    diskSize: Long,
+    externalBlockStoreSize: Long)
+
+@DeveloperApi
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockUpdatedInfo.scala]: Does this need to be made public? The user never has to create a BlockUpdateInfo object.

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+      blockManagers.remove(blockManagerId).foreach { blockIds =>
+        for (blockId <- blockIds) {
+          removeBlockFromBlockManager(blockId, blockManagerId)
+        }
+      }
+    }
+  }
+
+  private def removeBlockFromBlockManager(
+      blockId: BlockId, blockManagerId: BlockManagerId): Unit = {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: I think it is worth putting this method inside synchronized so that someone in future does not accidentally call this method without synchronizing.

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+
+    synchronized {
+      // Drop the update info if the block manager is not registered
+      blockManagers.get(blockManagerId).foreach { blocksInBlockManager =>
+        if (storageLevel.isValid) {
+          blocksInBlockManager.add(blockId)
+          val location = blockManagerId.hostPort
+          val newLocations =
+            blocks.get(blockId).map(_.locations).getOrElse(Set.empty) + location
+          val newStorageLevel = StorageLevel(
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: This set of lines seems duplicated in two methods. May be worth putting into a method to dedup. Also, it will make this method look less verbose.

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+            blockUIData.memSize,
+            blockUIData.diskSize,
+            blockUIData.externalBlockStoreSize,
+            newLocations))
+      }
+    }
+  }
+
+  def allBlocks: Seq[BlockUIData] = synchronized {
+    blocks.values.toBuffer
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: toBuffer or toSeq?

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+    blockId: BlockId,
+    storageLevel: StorageLevel,
+    memSize: Long,
+    diskSize: Long,
+    externalBlockStoreSize: Long,
+    locations: Set[String])
+
+private[spark] class BlockStatusListener extends SparkListener {
+
+  private val blockManagers = new mutable.HashMap[BlockManagerId, mutable.HashSet[BlockId]]
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: Why do you need to actually keep track of the block managers and corresponding blockIds in them? If all you need to provide to the UI is all block information (def allBlocks) then what is the need to keep track of block managers? You can just add or remove locations from the BlockUIData in the `blocks` map. Isnt it?

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+      blockManagers.get(blockManagerId).foreach { blocksInBlockManager =>
+        if (storageLevel.isValid) {
+          blocksInBlockManager.add(blockId)
+          val location = blockManagerId.hostPort
+          val newLocations =
+            blocks.get(blockId).map(_.locations).getOrElse(Set.empty) + location
+          val newStorageLevel = StorageLevel(
+            useDisk = diskSize > 0,
+            useMemory = memSize > 0,
+            useOffHeap = externalBlockStoreSize > 0,
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: Actually, isnt this effectively overwriting the previous storage level? What happens of a block was stored in memory is exec1 and then it gets an update that the exec2 adds it in disk. What should be the final storage level combining these two.

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: How about displaying like this:
screen shot 2015-07-02 at 10 18 44 pm

But it does not support sorting. If I click to sort some column, it will become like this:
screen shot 2015-07-02 at 10 18 52 pm

But I think the sorting feature is very important for this table, so I don't want to disable it. People usually want to find the biggest block.

Thoughts?

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala (hunk 19 above)
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: It would be helpful when a block manager is removed and we need to remove the blocks on this block manager.

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala (hunk 20 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: Yikes! Yeah we cant break sorting like that. Hmmm .. we need to think from scratch. Unlike RDDs, blocks can only be in one level per location. So showing "size in memory" AND "size in disk" does not make sense. It made sense for RDDs because some partitions may be in memory and some may be in disk. Let me thinks whats a better organization.

**Comment by zsxwing** [comment]
New screenshot:
screen shot 2015-07-06 at 10 34 24 pm

**Comment by tdas** [comment]
Jenkins test this again.

**Comment by tdas** [comment]
Jenkins, test this please.

**Comment by zsxwing** [comment]
retest this please

**Comment by tdas** [comment]
On the screenshot
- "Aggregated Stream Block Metrics by Executor" --> "Aggregated Metrics by executor" (Stream block part is obvious)
- I am not again categorizing the by Executor ID is a good idea. Rather I would like to see input blocks as the first column, sorted alphabetically. Then multiple rows per block ids as there are replicas. That would be easy to visually search through blocks of one input stream (order will show missing blocks). Also will be easy to see if any blocks has less replicas. 

I think the columns should be 
Block ID   |   Replication Level  |  Location  |  Storage Level  |   Size 

If replication level is 2, there will be two subrow with location + storage level + size each. The Storage Level will be either Memory, or Memory Serialized, or Disk, or External. 

Does that make sense.

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+    storageLevel: StorageLevel,
+    memSize: Long,
+    diskSize: Long,
+    externalBlockStoreSize: Long)
+
+/**
+ * The aggregated status of stream blocks in an executor
+ */
+private[spark] case class ExecutorStreamBlockStatus (
+  executorId: String,
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: incorrect formatting.

**Comment by tdas** [comment]
I see that you have changed the data structures to suit the desired table structure. But since the table structure has to change, i guess the data structures have to change. So I will not take a look at that right now.

**Comment by zsxwing** [comment]
Updated as per your suggestion:

g3

File: core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala [hunk 22]
```diff
@@ -0,0 +1,135 @@
...
+      blockManagerRemoved: SparkListenerBlockManagerRemoved): Unit = synchronized {
+    val blockManagerId = blockManagerRemoved.blockManagerId
+    blockManagers.get(blockManagerId).foreach { blocksInBlockManager =>
+      blocksInBlockManager.keys.foreach(
+        removeLocationFromBlockLocations(_, blockManagerId.hostPort))
+    }
+    blockManagers -= blockManagerId
+  }
+
+  def allExecutorStreamBlockStatus: Seq[ExecutorStreamBlockStatus] = synchronized {
```
Comment (tdas) [inline core/src/main/scala/org/apache/spark/storage/BlockStatusListener.scala]: These public functions can get called at different times leading to inconsistent values in the UI. Can you define a method that returns both information in a single synchronized call?

**Comment by zsxwing** [comment]
New screenshot:

g7

**Comment by andrewor14** [comment]
Screenshot looks great!

**Comment by tdas** [comment]
Everything else looks good, but I just realized that there are no UI tests! There should be tests added to StorageTabSuite, isnt it?

**Comment by zsxwing** [comment]
> Everything else looks good, but I just realized that there are no UI tests! There should be tests added to StorageTabSuite, isnt it?

I agree that we should do unit tests for UI stuffs. Let me see how to do some unit tests for `StorageTabSuite`. Actually I think many Spark UI tabs lack unit tests. Spark only has some `UISeleniumSuite`s to test UI. But `UISeleniumSuite` is actually an integration test rather than a unit test.

**Comment by zsxwing** [comment]
> Screenshot looks great! We can do it separately but maybe we should remove the RDD table if there are no RDD blocks cached.

Updated it in this PR.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by tdas** [comment]
Jenkins, test this please.

**Comment by tdas** [comment]
Jenkins, test this please.

**Comment by tdas** [comment]
LGTM, merging this!
